00X-tracing.md

```diff
@@ -0,0 +1,579 @@
+   *   If unspecified, then 'console' is used.
+   *
+   * NEW NODE.JS METHOD, EXISTING WEB METHOD + NEW OPTIONAL PARAMETERS FOR TRACING
+   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Console/timeStamp}
+   */
+  timeStamp(name: string, category?: string | string[]): void;
+
+  /**
+   * Writes a formatted message to the console with a stack trace.
+   *
+   * NOTE: This is not actually related to tracing, but is listed here because its name may
+   * cause some confusion.
+   *
+   * EXISTING NODE.JS & WEB METHOD (NO CHANGE)
+   * @see {@link https://nodejs.org/dist/latest-v7.x/docs/api/console.html#console_console_trace_message_args}
+   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Console/trace}
+   * @see {@link https://console.spec.whatwg.org/#trace}
+   */
+  trace(message?: any, ...optionalParams: any[]): void;
+
+  //
+  // Other non-tracing-related Console methods are omitted here.
+  //
+}
+```
+
+## 5. Implementation status
+### 5.1 Tracing fork
+Work is in progress at https://github.com/jasongin/nodejs/tree/tracejs. At the time of writing,
+tracing emitter APIs and console tracing API extensions are functional.
+
+The `CategoryEventEmitter` base class is implemented (with tests), however the full tracing
+listener implementation has a dependency on an "event callback" mode that is currently
+unimplemented in the v8 tracing code.
+
+### 5.2 Performance
+Performance for emitting tracing events from JavaScript is looking pretty good so far; I've done
+some basic optimization but not extensive tuning. On my dev system, a 5-year old desktop running
+Windows, the benchmark I wrote shows each single-category JS trace statement takes around **650ns
+with tracing enabled** (writing events to a JSON file), or around **45ns with tracing disabled**.
+Multi-category events are somewhat slower. Yes those numbers are nano-seconds, not micro- or
+milli-seconds.
+
+A more detailed performance analysis should be performed around the time of the pull request.
+
+## 6. Future work
+The following related items are outside the scope of this proposal, at least for now.
+
+### 6.1 Output serialization
+Currently, logs are simply JSON-serialized to a file; this JSON format is
+[understood by the Chrome dev tools](https://www.chromium.org/developers/how-tos/trace-event-profiling-tool). The C++ layer of the
+tracing framework supports pluggable exporters, for example Chromium has an
+[ETW exporter for Windows](https://chromium.googlesource.com/chromium/src/+/master/base/trace_event/trace_event_etw_export_win.h),
+which way may want to adapt to Node.js eventually. And there has been discussion of a more compact
+binary format. This is an area that needs some further development, however for performance reasons
+the tracing output probably would not be exposed to JavaScript anyway, beyond the listener APIs
+that are proposed here.
```

Comment:
Some time should be taken to design a processing pipeline for these. I think some tools and documentation to explain how to process this data format. Without interactive and easy-to-understand visualisations, tracing data is extremely hard to understand.

I should be straightforward to plot this data into a flamegraph for example.